# PCA在癌症基因组图谱数据中的应用
## 一、实验简介
### （一）问题描述
据统计，全球每年新增癌症患者达700万人，死于癌症的病人达500万人，60％的患者确诊后只能存活5年。目前已知的癌症有200多种，但是，无论什么癌症，在肿瘤的特殊类别（分型）或发展的不同分期方面都发现有基因组的特异变化，而正是基因组的改变（突变）导致了细胞分化、发育和生长通路的不正常，从而引发细胞不正常地失控增殖、生长。

通过应用基因组分析技术，特别是采用大规模的基因组测序，将人类全部癌症的基因组变异图谱绘制出来，并进行系统分析，旨在找到所有致癌和抑癌基因的微小变异，了解癌细胞发生、发展的机制，在此基础上取得新的诊断和治疗方法。

现如今，机器学习在数据分析领域已经慢慢走向成熟。如果可以用机器学习开发一套算法程序对癌症基因组图谱数据进行分析，辅助医生快速得出正确诊断结果，这将大大提高现代医学对癌症的诊断效率。

本实验借助了TCGA中的癌症基因组图谱数据，利用Sklearn机器学习库中的PCA算法对该数据集特征进行降维，然后利用SVM算法分别对降维前后的数据进行学习，最终使用这两个SVM算法模型对同一个新数据集的癌症种类进行对比分析。

### （二）问题需求
本次癌症基因组图谱数据集记录了801个样本数据，其特征数据是由癌症患者的RPPA【评估外显子组序列】 、DNA methylation【DNA甲基化】、Copy number【拷贝数变化(通过SNP阵列测量)】、Mutation【转录本剪接变化】、microRNA【microRNA表达】指标的数据构成，其中每个指标有多个数据，每个样本总共有20532个属性，样本标签是癌症的类型，数据集中总共记录了5种癌症类型，分别用数字1，2，3，4，5表示。

数据文件所在位置："pca\data.csv" 。   
本实验首先对该数据集按8：2的比例划分出了训练集和测试集，由于癌症基因组图谱数据集特征比较多，这对算法的训练和预测都造成了巨大困难，所以对训练集和测试集的数据特征进行了PCA降维，之后对降维前后的数据集分别训练同一个分类算法模型，让训练好的算法模型去预测测试集数据，最后对测试集数据结果进行准确率评估。

## 二、实验任务与要求
本实验调用Sklearn库中的PCA算法模块对癌症基因图谱泛癌症数据特征进行降维，用降维前后的数据分别来训练SVM算法模型，之后是用两个SVM模型对新的数据集进行预测分类，并对分类结果进行准确率评估，最后对PCA算法的降维效果进行分析。具体的实验要求如下：

1.观察数据：为了能初步了解数据特征的实际情况，可以观察前五条样本数据、每列数据的详细信息，以及数据中是否有空缺值。

2.数据处理：对观察数据阶段所提出的问题进行解决，之后将数据集按8:2划分为训练集和测试集，对训练集和测试集数据调用PCA算法进行降维并对该算法进行简单介绍。

3.PCA算法降维效果测试：使用降维前后的两个训练集分别对SVM算法模型进行训练，得到两个训练好的SVM算法模型对测试集数据进行预测，并对预测结果进行准确率评分，最后记录两个模型的训练、预测整个过程的时间。


## 三、实验步骤
本实验使用癌症基因组图谱数据，经过PCA算法模型对数据进行降维，之后供SVM算法模型进行学习与分类，比较PCA降维的效果，主要由以下几个步骤构成：

1.加载相关的包。

2.加载数据。

3.数据处理。

4.PCA算法降维效果测试。

### (一）加载相关的包
（1）加载numpy、pandas等包做数据处理。

（2）从sklearn.datasets和sklearn.model_selection中加载数据集和划分数据集。

（3）从sklearn.decomposition加载PCA算法模块对数据进行降维。

（4）从sklearn加载svm算法模块进行分类。

（5）从sklearn.metrics中加载accuracy_score评分的函数对模型分类结果计算准确率。

（6）加载time类记时。

In [14]:
import pandas as pd   #数据分析、处理包
import numpy as np   #矩阵、数组处理包
from sklearn.model_selection import train_test_split#数据划分函数
from sklearn.decomposition import PCA        #PCA降维算法 
from sklearn import svm   #支持向量机算法
from sklearn.metrics import accuracy_score   #准确率评分函数
import time      #时间戳模块

### (二）加载数据

In [15]:
#读取癌症基因组图谱数据特征
filename_data = 'pca\data.csv'
df_feature = pd.read_csv(filename_data)
#读取癌症基因组图谱数据标签
filename_label = 'pca\labels.csv'
df_label = pd.read_csv(filename_label)

### (三）数据处理

#### 1.观察数据

In [16]:
df_feature.head()   #观察数据集中的前5条

Unnamed: 0  gene_0    gene_1    gene_2    gene_3     gene_4  gene_5  \
0   sample_0     0.0  2.017209  3.265527  5.478487  10.431999     0.0   
1   sample_1     0.0  0.592732  1.588421  7.586157   9.623011     0.0   
2   sample_2     0.0  3.511759  4.327199  6.881787   9.870730     0.0   
3   sample_3     0.0  3.663618  4.507649  6.659068  10.196184     0.0   
4   sample_4     0.0  2.655741  2.821547  6.539454   9.738265     0.0   

     gene_6    gene_7  gene_8  ...  gene_20521  gene_20522  gene_20523  \
0  7.175175  0.591871     0.0  ...    4.926711    8.210257    9.723516   
1  6.816049  0.000000     0.0  ...    4.593372    7.323865    9.740931   
2  6.972130  0.452595     0.0  ...    5.125213    8.127123   10.908640   
3  7.843375  0.434882     0.0  ...    6.076566    8.792959   10.141520   
4  6.566967  0.360982     0.0  ...    5.996032    8.891425   10.373790   

   gene_20524  gene_20525  gene_20526  gene_20527  gene_20528  gene_20529  \
0    7.220030    9.119813   12.003135    9.650743    8.921326    5.286759   
1    6.256586    8.381612   12.674552   10.517059    9.397854    2.094168   
2    5.401607    9.911597    9.045255    9.788359   10.090470    1.683023   
3    8.942805    9.601208   11.392682    9.694814    9.684365    3.292001   
4    7.181162    9.846910   11.922439    9.217749    9.461191    5.110372   

   gene_20530  
0         0.0  
1         0.0  
2         0.0  
3         0.0  
4         0.0  

[5 rows x 20532 columns]

初步观察可知，数据集有20532列，其中第一列是样本的序号，不是样本特征，需要删除该列。

由于算法模型需要的全是数值型数据，在数据预处理前必须要对每列数据类型进行了解，可以利用info()函数来观察数据集中每列的数据类型。

In [17]:
df_feature.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 801 entries, 0 to 800
Columns: 20532 entries, Unnamed: 0 to gene_20530
dtypes: float64(20531), object(1)
memory usage: 125.5+ MB


从输出可以看到数据集所有列中只有一列是object类型，其余列都是float64类型，从上面所观察到数据集的前五列可知，该object类型是数据集的第一列。

由上表结合数据介绍可以了解到，数据特征比较多，数据情况比较复杂，而且都是数值型数据，若想要进一步了解这些数据，可以利用Pandas中的describe()函数对每列数据的总数量、平均值、标准差、最小值、下四分位数、中位数、上四分位数、最大值进行计算并显示出来。

In [18]:
describe = df_feature.describe()#观察数据的整体信息 ，注意数据的大小、行列规模、数据类型等，便于后续数据处理。  
describe

gene_0      gene_1      gene_2      gene_3      gene_4  gene_5  \
count  801.000000  801.000000  801.000000  801.000000  801.000000   801.0   
mean     0.026642    3.010909    3.095350    6.722305    9.813612     0.0   
std      0.136850    1.200828    1.065601    0.638819    0.506537     0.0   
min      0.000000    0.000000    0.000000    5.009284    8.435999     0.0   
25%      0.000000    2.299039    2.390365    6.303346    9.464466     0.0   
50%      0.000000    3.143687    3.127006    6.655893    9.791599     0.0   
75%      0.000000    3.883484    3.802534    7.038447   10.142324     0.0   
max      1.482332    6.237034    6.063484   10.129528   11.355621     0.0   

           gene_6      gene_7      gene_8      gene_9  ...  gene_20521  \
count  801.000000  801.000000  801.000000  801.000000  ...  801.000000   
mean     7.405509    0.499882    0.016744    0.013428  ...    5.896573   
std      1.108237    0.508799    0.133635    0.204722  ...    0.746399   
min      3.930747    0.000000    0.000000    0.000000  ...    2.853517   
25%      6.676042    0.000000    0.000000    0.000000  ...    5.454926   
50%      7.450114    0.443076    0.000000    0.000000  ...    5.972582   
75%      8.121984    0.789354    0.000000    0.000000  ...    6.411292   
max     10.718190    2.779008    1.785592    4.067604  ...    7.771054   

       gene_20522  gene_20523  gene_20524  gene_20525  gene_20526  gene_20527  \
count  801.000000  801.000000  801.000000  801.000000  801.000000  801.000000   
mean     8.765891   10.056252    4.847727    9.741987   11.742228   10.155271   
std      0.603176    0.379278    2.382728    0.533898    0.670371    0.580569   
min      6.678368    8.669456    0.000000    7.974942    9.045255    7.530141   
25%      8.383834    9.826027    3.130750    9.400747   11.315857    9.836525   
50%      8.784144   10.066385    5.444935    9.784524   11.749802   10.191207   
75%      9.147136   10.299025    6.637412   10.082269   12.177852   10.578561   
max     11.105431   11.318243    9.207495   11.811632   13.715361   11.675653   

       gene_20528  gene_20529  gene_20530  
count  801.000000  801.000000  801.000000  
mean     9.590726    5.528177    0.095411  
std      0.563849    2.073859    0.364529  
min      7.864533    0.593975    0.000000  
25%      9.244219    4.092385    0.000000  
50%      9.566511    5.218618    0.000000  
75%      9.917888    6.876382    0.000000  
max     12.813320   11.205836    5.254133  

[8 rows x 20531 columns]

从上表中可以看到数据集中有几列数据中的值为0.0的非常多，甚至列gene_5数据中全是0（无效值），这只是能看到的数据，还有很多列是不能看到的，在机器学习中无效值太多不仅会对算法模型的训练时间、预测时间有影响，而且还会使训练出的算法模型效果比较差，当然，这个问题可以用PCA算法来解决。

SVM算法模型对空缺值数据是比较敏感的，需要将空缺值检测出来并填充。由观察数据阶段可知，除了第一列数据类型其余列数据类型全是float64，如果出现空缺值只有np.nan这一种情况，下面利用Numpy库中的isnan()函数对数据类型为float64的列数据进行空缺值搜索，返回是否有np.nan的bool类型值。

In [19]:
np.any(np.isnan(df_feature.iloc[:,1:]))

False

这里可以看到输出的是False，说明数据集中没有空缺值。

#### 2.数据预处理

（1）删除非特征列  
根据前面步骤所分析的，第一列数据不是样本特征，为了降低训练难度，需要将这些非特征的列进行删除，可以利用Pandas库中的drop()函数从数据集中删除第一列数据。

In [20]:
df_feature = df_feature.drop([df_feature.columns[0]],axis=1)

（2）数据划分  
在不影响算法模型训练的情况下，使模型的评估更有可信度，可以将数据集按8：2划分为训练集和测试集，此处调用train_test_split(df_feature, df_label['Class_num'],test_size = 0.2)函数对数据进行划分。 

函数train_test_split()参数解释：

    第一个参数：样本特征数据；
    第二个参数：样本标签数据；
    第三个参数：训练集样本数据占总样本数据的比列；
    第四个参数：固定训练集和测试集的随机种子。

In [21]:
feature_train, feature_test, label_train, label_test = train_test_split(df_feature, df_label['Class_num'],test_size = 0.2,random_state=1)

（3）PCA降维   
PCA（主成分分析）是一种常用的无监督学习数据分析算法。该算法通过线性变换将原始数据变换为一组各维度线性无关的表示，可用于提取数据的主要特征分量，以最大化投影后的方差，其次是投影产生的损失最小为原则对数据集进行降维，常用于高维数据的降维。    
由于前面观察数据时发现了数据中有大量的0值，这对算法的训练并没有什么好处，可以利用PCA降维算法进行处理，首先调用Sklearn库中的PCA(n_components)类创建PCA算法类对象（参数n_components是降维后的维度），其次，通过对象调用fit_transform()方法对训练集数据进行拟合，最后将拟合好的模型对数据进行降维。n_components可以设置一个较小的值，然后根据下面检验是否合理再不断增大该值。函数参数解释：

    第一个参数：需要降维的特征数据。
调用transform()方法对测试集数据进行降维。该函数参数解释：
    
    第一个参数：需要降维的特征数据。
注意transform()和fit_transform()函数返回的是数组，在下面索引行和索引列时需注意语法，数组中没有.iloc。

In [22]:
n_components=50      #设置降维后的维度
pca = PCA(n_components)    #实例化PCA降维类
feature_train_pca = pca.fit_transform(feature_train)  #对训练集数据进行PCA降维
feature_test_pca = pca.transform(feature_test)  #对测试集数据进行PCA降维

检验降维是否合理：   
PCA算法是根据方差来降维的，因此可以利用降维后各成分方差情况来判断降维后的维数是否合理。PCA算法类的属性explained_variance_ratio_是降维后的各成分方差占比，比例越大，保留的信息越多，此处计算降维后各成分的方差占比之和，一般该值在70-95之间，从下面结果可看出样本数据降到50维是可以的。

In [27]:
sum(pca.explained_variance_ratio_)

0.7128434151978347

### (四）PCA算法降维效果测试

支持向量机（SVM）是一种二类分类模型，其基本模型定义为特征空间上的间隔最大的线性分类器，其学习策略是使间隔最大化，最终可转化为一个凸二次规划问题的求解。该算法不仅适用于线性问题，还适用于非线性问题(用核函数)，由于本次数据特征比较多，降维后数据间将不存在线性关系，因此这里选用SVM进行PCA算法来对PCA降维的效果进行测试。   

测试PCA降维效果需要使用降维前后的两个训练集分别对SVM算法模型进行训练，得到两个训练好的SVM算法模型对测试集数据进行预测，并对预测结果进行准确率评分。为了避免代码重复，此处编写一个函数test_model(feature_train, feature_test, label_train, label_test)实现SVM算法模型的调用、训练、预测以及预测结果评估，并对该过程的时间进行了记录，函数参数解释:

    feature_train：训练集特征数据；
    feature_test：测试集特征数据；
    label_train：训练集标签数据；
    label_test：测试集标签数。

In [32]:
def test_model(feature_train, feature_test, label_train, label_test):
    starttime = time.time()
    model = svm.SVC(kernel='sigmoid')
    model.fit(feature_train, label_train)
    #make prediction of the gene dataset 
    label_prediction = model.predict(feature_test)
    print('预测结果的准确率：',accuracy_score(label_test,label_prediction),'耗时:',time.time()-starttime)

使用降维前的数据集对模型进行训练。

In [33]:
test_model(feature_train,feature_test,label_train,label_test)

E:\profession_install\Anaconda\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


预测结果的准确率： 0.3416149068322981 耗时: 17.94297981262207


使用降维后的数据集对模型进行训练。

In [34]:
test_model(feature_train_pca,feature_test_pca,label_train,label_test)

预测结果的准确率： 0.8944099378881988 耗时: 0.015955448150634766


E:\profession_install\Anaconda\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


## 四、结果分析

本实验采用癌症基因组图谱数据进行PCA降维，使用降维前后的数据分别训练SVM算法模型，并评估模型对测试集数据分类的准确率。从上面的实验中可以看出：未进行PCA降维的数据训练出的SVM算法模型预测测试集，其预测结果的准确率只有34.2%，整个过程耗时达到17秒多；然而，经过降维之后的数据训练同一个模型对测试集进行预测，其预测结果可达到了90%左右的准确率，并且耗时仅需0.015秒。得到这个结果的可能原因是：首先，说明特征数量对训练和预测的时间会有明显的影响，特征数量越多其耗时越久；其次说明数据集中的无效值多，引起数据分类困难，使训练模型和模型预测的效果都会变差。而PCA算法以最大化投影后的方差，投影产生的损失最小为原则对数据集进行降维，这样不仅降低了数据分类的困难，而且还降低了数据的特征数量，因此进行PCA降维后的数据在模型的训练时间和预测结果上都要比未经降维的数据要优。

In [2]:
from sklearn.decomposition import PCA
from sklearn.datasets import load_iris


In [3]:
data = load_iris()

In [5]:
x = data.data
x

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [6]:
model = PCA(n_components=3)

In [7]:
model.fit(X=x)#训练模型

PCA(n_components=3)

In [9]:
model.fit_transform(x)#降维后的数据

array([[-2.68412563,  0.31939725, -0.02791483],
       [-2.71414169, -0.17700123, -0.21046427],
       [-2.88899057, -0.14494943,  0.01790026],
       [-2.74534286, -0.31829898,  0.03155937],
       [-2.72871654,  0.32675451,  0.09007924],
       [-2.28085963,  0.74133045,  0.16867766],
       [-2.82053775, -0.08946138,  0.25789216],
       [-2.62614497,  0.16338496, -0.02187932],
       [-2.88638273, -0.57831175,  0.02075957],
       [-2.6727558 , -0.11377425, -0.19763272],
       [-2.50694709,  0.6450689 , -0.07531801],
       [-2.61275523,  0.01472994,  0.10215026],
       [-2.78610927, -0.235112  , -0.20684443],
       [-3.22380374, -0.51139459,  0.06129967],
       [-2.64475039,  1.17876464, -0.15162752],
       [-2.38603903,  1.33806233,  0.2777769 ],
       [-2.62352788,  0.81067951,  0.13818323],
       [-2.64829671,  0.31184914,  0.02666832],
       [-2.19982032,  0.87283904, -0.12030552],
       [-2.5879864 ,  0.51356031,  0.21366517],
       [-2.31025622,  0.39134594, -0.239